### ISSUE TITLE
second derivatives not support

### ISSUE THREAD

**Issue by Soooooda69** [NVlabs/nvdiffrast#218]
Hi, I'm trying to integrate the differentiable renderer as part of my DL model. When I run back propagations it works fine, however, when I do second order derivatives, it seems the computation graph breaks at the differentiable renderer, making the gradients failed to propagate from output of the model to the input. I wonder is this commonly known for the renderer or is it just bug on my end, appreciate it!

Here's my renderer implementation attached:

```
class DepthRenderer(nn.Module):
    def __init__(self, cfg):
        super().__init__()
        self.znear = cfg.cam.znear
        self.zfar = cfg.cam.zfar
        self.resolution = [cfg.cam.W, cfg.cam.H]
        self.glctx = dr.RasterizeCudaContext(device=cfg.device)
        self.device = cfg.device
        self.intrinsics = cfg.cam.intrinsics
        self.setup(os.path.join(cfg.data_root, 'localMedium.ply'))

    def setup(self, geometry_path):
        # load geometry
        geometry = o3d.io.read_triangle_mesh(geometry_path)
        normals = np.asarray(geometry.vertex_normals)
        normals = (1-normals) / 2
  
        # Convert Open3D mesh to PyTorch3D mesh
        self._pos = torch.tensor(np.asarray(geometry.vertices), dtype=torch.float32).to(self.device)
        self._pos_idx = torch.tensor(np.asarray(geometry.triangles), dtype=torch.int32).to(self.device)
        self._col_idx = self._pos_idx
        self._col = torch.tensor(np.asarray(geometry.vertex_colors), dtype=torch.float32).to(self.device)
        self._normals = torch.tensor(normals, dtype=torch.float32).to(self.device)
        print("Mesh has %d triangles and %d vertices." % (self._pos_idx.shape[0], self._pos.shape[0]))
        # get projection matrix
        glproj = glprojection(K=self.intrinsics, w=self.resolution[0], h=self.resolution[1], n=self.znear, f=self.zfar)
        self.glproj = torch.tensor(glproj,
                              dtype=torch.float32).to(self.device)

        self.convert_rotation = torch.eye(4, device=self.device)
        rot_mat = R.from_euler('xyz', [180, 0, 0], degrees=True).as_matrix()
        self.convert_rotation[:3, :3] = torch.tensor(rot_mat, dtype=torch.float32).to(self.device)
    
    def transform_pos(self, mtx, pos):
        """
        Transforms a batch of positions using a batch of transformation matrices.

        Args:
            mtx (torch.Tensor or np.ndarray): Transformation matrix of shape [B, 4, 4] or [4, 4].
            pos (torch.Tensor): Positions of shape [N, 3].

        Returns:
            torch.Tensor: Transformed positions of shape [B, N, 3].
        """
        # Convert numpy to torch and ensure correct device
        t_mtx = torch.from_numpy(mtx).to(pos.device) if isinstance(mtx, np.ndarray) else mtx
        
        # Ensure transformation matrix has batch dimension
        if t_mtx.ndim == 2:  # If single [4,4] matrix, expand to batch size
            t_mtx = t_mtx.unsqueeze(0)

        # Convert (x, y, z) -> (x, y, z, 1) to get homogeneous coordinates
        ones = torch.ones((pos.shape[0], 1), device=pos.device)  # Shape: [N, 1]
        posw = torch.cat([pos, ones], dim=-1)  # Shape: [N, 4]

        # Apply transformation to each batch
        transformed = torch.matmul(posw.unsqueeze(0), t_mtx.transpose(1, 2))  # [B, N, 4]
        return transformed

    def render(self, mtx):
        pos_clip    = self.transform_pos(mtx, self._pos)
        depth_ = pos_clip[..., 2:3].contiguous()
        rast_out, _ = dr.rasterize(self.glctx, pos_clip, self._pos_idx, resolution=self.resolution)
        depth_i, _ = dr.interpolate(depth_, rast_out, self._pos_idx)
        depth    = dr.antialias(depth_i, rast_out, pos_clip, self._pos_idx)
        return depth, depth_, depth_i, rast_out, pos_clip
    
    def forward(self, extrinsics, opt_rel=None):
        if opt_rel is None:
            opt_rel = torch.eye(4, device=extrinsics.device)
        w2cs = torch.linalg.inv(extrinsics @ opt_rel)
        gl_w2cs = torch.matmul(self.convert_rotation, w2cs)
        # gl_w2cs = w2cs
        mvp = torch.matmul(self.glproj, gl_w2cs)
        depth, depth_, depth_i, rast_out, pos_clip = self.render(mvp)
        return depth.permute(0, 3, 1, 2)
```

**Comment by Soooooda69** [comment]
Thanks for the clarification. Do you have any plans for supporting the second-order derivative of the renderer, if not, is there any recommendations, appreciate it!

**Comment by s-laine** [comment]
There are no plans to add support for higher-order derivatives. I don't know if there are any renderers that support them.
